### PR TITLE
Merges Master and Greens

### PR DESCRIPTION
**Pull request by orta**
There were three main problems stopping us WRT tests:

* `AREmission sharedEmission` was not set up.
* ARTopMenuViewController would occasionally create ARHomeComponentVCs that would start to do their own React-y work without us expecting that to happen, which causes main thread work to happen in other tests.
* Some logic around multiple taps on the tab in the nav would crash.

Will comment on these three bits inline, the rest is all from Ash

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>

    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">Release 2.6.4 does not have a date</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>

    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Needs testing on a Phone if change is non-trivial</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Detected some Swift building time outliers</td>
    </tr>
  </tbody>
</table>

Time | Class | Function |
| --- | ----- | ----- |243.7ms | [AuctionViewController.swift](/artsy/eigen/tree/orta-v3-green/Artsy/View_Controllers/Auction/AuctionViewController.swift#L199) | @objc dynamic func showRefineTappedAnimated(animated: Bool)
201.4ms | [WebSocket.swift](/artsy/eigen/tree/orta-v3-green/Pods/Starscream/Source/WebSocket.swift#L528) | private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
195.6ms | [SwiftyJSON.swift](/artsy/eigen/tree/orta-v3-green/Pods/SwiftyJSON/Source/SwiftyJSON.swift#L445) | get {}
87.9ms | [LiveAuctionLotViewController.swift](/artsy/eigen/tree/orta-v3-green/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift#L47) | @objc override func viewDidLoad()
84.9ms | [Observable.swift](/artsy/eigen/tree/orta-v3-green/Pods/Interstellar/Sources/Observable.swift#L71) | private final func nextTokenHash() -> Int
71.7ms | [AuctionLotMetadataStackScrollView.swift](/artsy/eigen/tree/orta-v3-green/Artsy/View_Controllers/Live_Auctions/Views/AuctionLotMetadataStackScrollView.swift#L18) | required init(viewModel: LiveAuctionLotViewModelType, salesPerson: LiveAuctionsSalesPersonType, sideMargin: String)
71.4ms | [LiveAuctionsAdminViewController.swift](/artsy/eigen/tree/orta-v3-green/Artsy/View_Controllers/Live_Auctions/LiveAuctionsAdminViewController.swift#L22) | @objc override func viewDidLoad()
65.9ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/orta-v3-green/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L25) | func ar_presentBlurredOverlayWithTitle(title: String, subtitle: String, buttonState: BlurredStatusOverlayViewCloseButtonState = default)
63.9ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/orta-v3-green/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L31) | (closure)
58.4ms | [RefinementOptionsViewController+Private.swift](/artsy/eigen/tree/orta-v3-green/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift#L63) | func stackView() -> ORStackView


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m [hunk 1]
```diff
@@ -465,25 +469,34 @@ - (BOOL)tabContentView:(ARTabContentView *)tabContentView shouldChangeToIndex:(N
     if (index == self.selectedTabIndex) {
         ARNavigationController *controller = (id)[tabContentView currentNavigationController];
 
-        if (controller.viewControllers.count == 1) {
-            UIScrollView *scrollView = nil;
-            if (index == ARTopTabControllerIndexFeed) {
-                scrollView = [(ARSimpleShowFeedViewController *)[controller.childViewControllers objectAtIndex:0] tableView];
```
Comment (orta) [inline Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m]: Assumed an `ARHomeComponent`  was a `ARSimpleShowFeedViewController` which caused a crash, which failed a test

File: Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m [hunk 2]
```diff
@@ -465,25 +469,34 @@ - (BOOL)tabContentView:(ARTabContentView *)tabContentView shouldChangeToIndex:(N
...
 
+        // Otherwise find the first scrollview and pop to top
+        else if (index == ARTopTabControllerIndexFeed ||
+            index == ARTopTabControllerIndexBrowse ||
+            index == ARTopTabControllerIndexFavorites) {
+
+            UIViewController *currentRootViewController = [controller.childViewControllers first];
+            UIScrollView *rootScrollView = (id)[self firstScrollToTopScrollViewFromRootView:currentRootViewController.view];
+            [rootScrollView setContentOffset:CGPointMake(rootScrollView.contentOffset.x, -rootScrollView.contentInset.top) animated:YES];
+        }
```
Comment (orta) [inline Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m]: this now finds the first correct scrollview and uses that instead, is null safe too, so if React native makes their own ScrollView then we will only lose that feature, not break

File: Artsy_Tests/View_Controller_Tests/App_Menu/ARTestTopMenuNavigationDataSource.h [hunk 3]
```diff
@@ -1,9 +1,20 @@
...
 @end
+
+/// A data source that replicates the API almost exactly, however it will never
+/// return the actual view controllers inside it, only new UIViewControllers
+
+/// This is the default for -ARTopMenuViewController initWithStubbedViewControllers
+
+@interface ARStubbedTopMenuNavigationDataSource : ARTopMenuNavigationDataSource
+
+@end
```
Comment (orta) [inline Artsy_Tests/View_Controller_Tests/App_Menu/ARTestTopMenuNavigationDataSource.h]: this is the new addition, it won't return the actual view controllers

File: Artsy_Tests/View_Controller_Tests/App_Menu/ARTopMenuViewControllerSpec.m [hunk 4]
```diff
@@ -185,7 +185,7 @@ - (JSBadgeView *)badgeForButtonAtIndex:(NSInteger)index createIfNecessary:(BOOL)
 describe(@"navigation", ^{
    __block NSInteger tabIndex;
    before(^{
-       dataSource = [[ARTopMenuNavigationDataSource alloc] init];
+       dataSource = [[ARStubbedTopMenuNavigationDataSource alloc] init];
        sharedBefore();
```
Comment (orta) [inline Artsy_Tests/View_Controller_Tests/App_Menu/ARTopMenuViewControllerSpec.m]: and this was the test that caused the trouble with a LiveAuction async one further down the line

File: Pods [hunk 5]
```diff
@@ -1 +1 @@
-Subproject commit 737a51f0f819e1cc39c42081c15712681e7d593d
+Subproject commit e417bdfccf207d874ebec11c03ab0534ddb95892
```
Comment (orta) [inline Pods]: all of ash's stuff, and a bit of mine

**Comment by orta** [comment]
hah - nice one Danger 

> Release 2.6.4 does not have a date

File: Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m [hunk 6]
```diff
@@ -468,7 +471,16 @@ - (BOOL)tabContentView:(ARTabContentView *)tabContentView shouldChangeToIndex:(N
         if (controller.viewControllers.count == 1) {
             UIScrollView *scrollView = nil;
             if (index == ARTopTabControllerIndexFeed) {
-                scrollView = [(ARSimpleShowFeedViewController *)[controller.childViewControllers objectAtIndex:0] tableView];
+                ARHomeComponentViewController *homeComponentVC = [controller.childViewControllers objectAtIndex:0];
+                UIView *rootView = homeComponentVC.view;
+                while (rootView.subviews.firstObject && ![rootView isKindOfClass:RCTScrollView.class]) {
+                    rootView = rootView.subviews.firstObject;
+                    NSLog(@"Root:  %@", rootView);
+                }
```
Comment (alloy) [inline Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m]: Can we move this and [that](https://github.com/artsy/eigen/blob/4454d6bb595a4fcc0d491c75eed07e24cae52300/Artsy/App/ARAppDelegate%2BEmission.m#L161-L174) into the `ARComponentViewController` in Emission and use it from both places?

**Comment by alloy** [comment]
Nice work 👏

File: Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m (hunk 6 above)
Comment (orta) [inline Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m]: hah, I knew there was something like that

**Pull request merged**